### PR TITLE
Add automatic generation of conda binary packages for Windows, macOS and Linux in robotology channel

### PR DESCRIPTION
**Pull request by traversaro**
This PR fixes https://github.com/robotology/robotology-superbuild/issues/620 . In particular, it adds machinery to generate the **conda recipes for each subproject of the superbuild**, and a GitHub Action job that build binaries once a week for all the generated recipes for the latest release of the robotology-superbuild projects and loads them in the [**robotology conda channel**](https://anaconda.org/robotology), that builds on top of the [`conda-forge`](https://conda-forge.org/) similarly to who the [`robostack`](https://github.com/RoboStack/ros-noetic) channel for ROS software build on top of [`conda-forge`](https://conda-forge.org/) as well.

It is unfortunately a relatively large PR, but I preferred to experiment with it a bit to understand there is something functional before adding it to the official robotology-superbuild repo. In particular, the test binaries were generated in the channel https://anaconda.org/robotology-staging . The modifications can be largely divided in 5 main categories:
1. Changes in the `conda` directory, that contain all the machinery required to generate conda recipes out for all superbuild projects. These internal mechanism are described in [`doc/conda-recipe-generation.md`](https://github.com/robotology/robotology-superbuild/pull/652/files#diff-1448456855abb123feefa353fc9f53bd6da5f2a085c55da7107bfa3fa554e947) document
2. The GitHub Action generate-conda-packages ( `.github/workflows/generate-conda-packages.yaml` ) that uses the machinery described before to actually compile the conda binaries and upload them to [**robotology conda channel**](https://anaconda.org/robotology)
3. Modifications in the `cmake/Build***.cmake`  scripts, that are basically just adding metadata required to generate conda packages 
4. Modification in developer-oriented documentation in ` doc/developers-faqs.md` (https://github.com/robotology/robotology-superbuild/pull/652/files#diff-f9b520063217385eaf8b07ab7805449c26b59ccb1158241d51497bff72e92100) that are oriented towards package and profile mantainers describing how to add the required conda metadata when they add a new package
5. Modification in user-oriented documentation in `doc/conda-forge.md` ( https://github.com/robotology/robotology-superbuild/pull/652/files#diff-58f523b9fc5050a95da9ce381d330a2cf3ad7f1c88e6fb407fad9fd87d09d091 ): this are meant for end-users of robotology software that want to install the robotology packages directly in binary format. As this is still experimental and the `robotology` channel is still empty, this documentation is still **not linked** in the main README.

Given the size of the PR, I think it would be ideal if reviewer could focus their attention to points 4. and 5. , while the first three points are mainly internal implementation details.

### CODE REVIEW

**Comment by traversaro** [comment]
As just a bunch of technical details is not a fun, this is a quick peek to what can be done now that we have (obviously all the instructions use `robotology-staging` until this PR is merged) 

## Use robotology-superbuild software built with ROBOTOLOGY_USES_MATLAB in MATLAB online 

If you are a user of [MATLAB online](https://matlab.mathworks.com/), you may want to run your scripts directly on the cloud, but this could be problematic if you depend on some MATLAB or Simulink libraries provided by the robotology-superbuild, given that `MATLAB online` does not permit you to install build tools packages via apt or other system package managers. Thanks to conda, you can easily install MATLAB libraries such as the [YARP Matlab bindings](https://github.com/robotology/yarp-matlab-bindings) of the robotology-superbuild simple with these MATLAB commands:

~~~
system('curl -LO https://github.com/conda-forge/miniforge/releases/latest/download/Miniforge3-Linux-x86_64.sh')
system('sh Miniforge3-Linux-x86_64.sh -b')
system('~/miniforge3/condabin/conda install -c robotology-staging yarp-matlab-bindings -y')
addpath('~/miniforge3/mex')
~~~
yarp_bindings_matlab_online

(At this moment `idyntree` MATLAB bindings  and `wb-toolbox` do not work due to https://github.com/robotology/robotology-superbuild/issues/653, but this should be hopefully easy to fix).


## Quickly install software on Windows

While the process of installing the robotology-superbuild and its dependencies has been historically relatively smooth on Ubuntu and macOS, it has always been quite complex on Windows. The typical pattern indeed was of Windows user that had to run Docker, a Virtual Machine or WSL just to install it. The conda binaries drastically simplify the process of installing robotology packages on Windows, for example if you have conda installed (see instructions in https://github.com/robotology/robotology-superbuild/blob/32fa9c33fb597b83aef936528e2d476576e96a5d/doc/install-miniforge.md) you just need to run the following commands:
~~~
conda create -n robsubenv
conda activate robsubenv
conda install -c robotology-staging icub-models gazebo-yarp-plugins
~~~

to be able to run Gazebo and an iCub simulation:
robsubenv


## Easily run Continuous Integration of C++ libraries that depend on robotology libraries

Thanks to the conda binaries provided by the `robotology` channel, you can easily setup a continuous integration setup using GitHub Actions that runs on Windows, macOS and Linux without having to manually compile your dependencies or manually handling cache. For example, this is the script that is necessary to run CI for [`bipedal-locomotion-framework`](https://github.com/dic-iit/bipedal-locomotion-framework):

~~~yaml
name: C++ CI Workflow with conda-forge dependencies

on:
  push:
  pull_request:
  schedule:
  # * is a special character in YAML so you have to quote this string
  # Execute a "nightly" build at 2 AM UTC 
  - cron:  '0 2 * * *'

jobs:
  build:
    name: '[${{ matrix.os }}@${{ matrix.build_type }}@conda]'
    runs-on: ${{ matrix.os }}
    strategy:
      matrix:
        build_type: [Release]
        os: [ubuntu-latest, windows-latest, macOS-latest]
      fail-fast: false

    steps:
    - uses: actions/checkout@v2

    - uses: conda-incubator/setup-miniconda@v2
      with:
        mamba-version: "*"
        channels: conda-forge,defaults
        channel-priority: true

    - name: Dependencies
      shell: bash -l {0}
      run: |
        # Compilation related dependencies 
        mamba install cmake compilers make ninja pkg-config
        # Actual dependencies
        mamba install -c robotology-staging idyntree yarp matio-cpp lie-group-controllers eigen qhull casadi cppad manif spdlog catch2
        
    - name: Configure [Linux&macOS]
      if: contains(matrix.os, 'macos') || contains(matrix.os, 'ubuntu')
      shell: bash -l {0}
      run: |
        mkdir -p build
        cd build
        cmake -GNinja -DBUILD_TESTING:BOOL=ON -DCMAKE_BUILD_TYPE=${{ matrix.build_type }} ..
      
    - name: Configure [Windows]
      if: contains(matrix.os, 'windows')
      shell: bash -l {0}
      run: |
        mkdir -p build
        cd build
        cmake -G"Visual Studio 16 2019" -DBUILD_TESTING:BOOL=ON -DCMAKE_BUILD_TYPE=${{ matrix.build_type }} ..

    - name: Build 
      shell: bash -l {0}
      run: |
        cd build
        cmake --build . --config ${{ matrix.build_type }}
        
    - name: Test
      shell: bash -l {0}
      run: |
        cd build
        ctest --output-on-failure -C ${{ matrix.build_type }}
~~~

If you have a CMake project, the only line that you will need to modify to adapt it to your project is just the line:
~~~
mamba install -c robotology-staging idyntree yarp matio-cpp lie-group-controllers eigen qhull casadi cppad manif spdlog catch2
~~~

using your dependencies instead of the one of `bipedal-locomotion-framework` .

**Comment by traversaro** [comment]
I also plan to eventually do in-person training on this new Conda-based tools to group of people, but before reaching that point I would prefer to have all the package generation machinery up and running.

**Comment by traversaro** [comment]
cc @Tobias-Fischer

File: cmake/BuildICUB.cmake
Comment (Tobias-Fischer) [inline cmake/BuildICUB.cmake:95]: I've often seen "UNIX AND NOT APPLE" - not sure which one is better.

File: cmake/BuildYARP.cmake
Comment (Tobias-Fischer) [inline cmake/BuildYARP.cmake:8]: Out of interest, doesn't that break non-conda builds?

File: conda/conda_build_config.yml
Comment (Tobias-Fischer) [inline conda/conda_build_config.yml:2]: If you use this, make sure to add the `__osx` run dependency: https://conda-forge.org/docs/maintainer/knowledge_base.html

File: conda/python/generate_conda_recipes_from_metadatadata.py
Comment (Tobias-Fischer) [inline conda/python/generate_conda_recipes_from_metadatadata.py:1]: Mistake in filename? metadatadata?

File: conda/recipe_template/meta.yaml
Comment (Tobias-Fischer) [inline conda/recipe_template/meta.yaml:45]: Add the osx run stuff here I guess

File: doc/conda-forge.md
Comment (Tobias-Fischer) [inline doc/conda-forge.md:72]: Typo installl

File: doc/conda-recipe-generation.md
Comment (Tobias-Fischer) [inline doc/conda-recipe-generation.md:7]: Ran out of energy there? :)

File: doc/developers-faqs.md
Comment (Tobias-Fischer) [inline doc/developers-faqs.md:75]: ofter -> often

**Comment by Tobias-Fischer** [comment]
Amazing work, well done @traversaro. This is super exciting :)

File: cmake/BuildYARP.cmake
Comment (traversaro) [reply cmake/BuildYARP.cmake:8]: `ACE`  needs to be installed in the system on all platforms supported by the robotology-superbuild. In general, in the superbuild build files there were a few of these `find_package`  for dependencies that are not handled by the superbuild, but to avoid problems in the conda recipe generation process I just removed them for now, so that eventually we can add them back consistently for all packages.

File: conda/recipe_template/meta.yaml
Comment (traversaro) [reply conda/recipe_template/meta.yaml:45]: Done in https://github.com/robotology/robotology-superbuild/pull/652/commits/05593fddf73fac940540d6d48e1b181e11b7351d .

File: conda/python/generate_conda_recipes_from_metadatadata.py
Comment (traversaro) [reply conda/python/generate_conda_recipes_from_metadatadata.py:1]: Good catch.

File: conda/conda_build_config.yml
Comment (traversaro) [reply conda/conda_build_config.yml]: Done in https://github.com/robotology/robotology-superbuild/pull/652/commits/05593fddf73fac940540d6d48e1b181e11b7351d .

File: doc/conda-recipe-generation.md
Comment (traversaro) [reply doc/conda-recipe-generation.md]: Yes. Fixed in https://github.com/robotology/robotology-superbuild/pull/652/commits/f0c615e0a03de7466ff4253b55d23730ec29ad84 .

File: doc/developers-faqs.md
Comment (traversaro) [reply doc/developers-faqs.md]: Fixed in https://github.com/robotology/robotology-superbuild/pull/652/commits/cfe30cd2c9ff38a17e589b371d830634f0417c47 .

**Comment by traversaro** [comment]
> 
> 
> Amazing work, well done @traversaro. This is super exciting :)

Thanks a lot for the feedback @Tobias-Fischer ! I should have fixed all of them.

File: cmake/BuildICUB.cmake
Comment (traversaro) [reply cmake/BuildICUB.cmake:95]: `UNIX AND NOT APPLE` is typically used to include *bsd systems, as conda-forge do not support any *bsd or similar target at the moment I think it is more clear to indicate it as "do something if we are on linux". If conda-forge will ever support other non-Linux targets (such as WebAssembly ( https://github.com/iodide-project/pyodide/issues/795 ), that would be a dream!) we will eventually fix this.

File: doc/conda-forge.md
Comment (nunoguedelha) [inline doc/conda-forge.md:73]:
```suggestion
If you want to develop some C++ code on the top of these libraries, it is recommended to also install the necessary compiler and development tools directly in the same environment:
```

File: doc/conda-forge.md
Comment (nunoguedelha) [inline doc/conda-forge.md:68]:
```suggestion
In addition, if you want to simulate the iCub in Gazebo, you should also install `icub-models` and `gazebo-yarp-plugins`:
```

File: doc/developers-faqs.md
Comment (nunoguedelha) [inline doc/developers-faqs.md:72]:
```suggestion
* If your package needs to set or modify some environment variables to work correctly, it should provide a pair of [multisheller](https://github.com/wolfv/multisheller) scripts named `<condaPkgName>_activate.msh` and `<condaPkgName>_deactivate.msh` in the `conda/multisheller` directory to describe how the environment should be modified. Refer to the existing scripts for more details.
```

**Review by nunoguedelha — CHANGES_REQUESTED** [review]

**Review by nunoguedelha — APPROVED** [review]

**Comment by nunoguedelha** [comment]
Sorry @traversaro , there is a minor typo I forgot to mention, and actually is not among the modified files. It's in https://github.com/robotology/robotology-superbuild/blob/master/doc/install-miniforge.md#install-1:
```
# Download
curl -fsSLo Miniforge3.sh https://github.com/conda-forge/miniforge/releases/latest/download/Miniforge3-MacOSX-$(uname -m).sh
# Install with default options
bash ./Miniforge3-MacOSX-$(uname -m).sh
```
Second line should be `bash ./Miniforge3.sh`.

**Comment by traversaro** [comment]
> 
> 
> Sorry @traversaro , there is a minor typo I forgot to mention, and actually is not among the modified files. It's in https://github.com/robotology/robotology-superbuild/blob/master/doc/install-miniforge.md#install-1:
> 
> ```
> # Download
> curl -fsSLo Miniforge3.sh https://github.com/conda-forge/miniforge/releases/latest/download/Miniforge3-MacOSX-$(uname -m).sh
> # Install with default options
> bash ./Miniforge3-MacOSX-$(uname -m).sh
> ```
> 
> Second line should be `bash ./Miniforge3.sh`.

Fix for this proposed in https://github.com/robotology/robotology-superbuild/pull/664 .

**Comment by traversaro** [comment]
Thanks @nunoguedelha !

**Pull request merged**
